p2p/src/transports/mconnection.rs

```diff
@@ -0,0 +1,319 @@
+
+use super::super::secret_connection::{SecretConnection, Version};
+use super::super::transport::*;
+
+/// A Transport which multiplexes messages from different sources over a single TCP stream.
+/// Spec: https://github.com/tendermint/spec/blob/master/spec/p2p/connection.md#p2p-multiplex-connection
+pub struct MConnectionTransport {}
+
+/// A `TcpStream` wrapped in `SecretConnection`.
+pub struct MConnection {
+    public_key: ed25519::PublicKey,
+    secret_connection: SecretConnection<TcpStream>,
+    stream: TcpStream,
+}
+
+/// An `Endpoint` for connecting to other peers.
+pub struct MEndpoint {
+    private_key: Arc<ed25519::Keypair>,
+    protocol_version: Version,
+    listen_addrs: Vec<SocketAddr>,
+}
+
+/// An `Iterator` for accepting connections.
+pub struct MIncoming {
+    tcp_listener: TcpListener,
+    private_key: Arc<ed25519::Keypair>,
+    protocol_version: Version,
+}
+
+impl MConnection {
+    /// Opens a TCP connection to a remote host and establishes a secret connection
+    /// [`SecretConnection`].
+    pub fn connect(
+        addr: &SocketAddr,
+        private_key: &ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<Self> {
+        let stream = TcpStream::connect(addr)?;
+        let stream_clone = stream.try_clone()?;
+
+        let public_key = private_key.public;
+        let secret_connection = SecretConnection::new(stream, &private_key, protocol_version)?;
+
+        Ok(Self {
+            public_key,
+            secret_connection,
+            stream: stream_clone,
+        })
+    }
+
+    /// Opens a TCP connection to a remote host with a timeout and establishes a secret connection
+    /// [`SecretConnection`].
+    pub fn connect_timeout(
+        addr: &SocketAddr,
+        timeout: Duration,
+        private_key: &ed25519::Keypair,
+        protocol_version: Version,
+    ) -> Result<Self> {
+        let stream = TcpStream::connect_timeout(addr, timeout)?;
+        let stream_clone = stream.try_clone()?;
+
+        let public_key = private_key.public;
+        let secret_connection = SecretConnection::new(stream, &private_key, protocol_version)?;
+
+        Ok(Self {
+            public_key,
+            secret_connection,
+            stream: stream_clone,
+        })
+    }
+}
+
+impl Connection for MConnection {
+    type Error = std::io::Error;
+    type Read = SecretConnection<TcpStream>;
+    type Write = SecretConnection<TcpStream>;
+
+    fn close(&self) -> Result<()> {
+        self.stream
+            .shutdown(Shutdown::Both)
+            .wrap_err("failed to shutdown")
+    }
+
+    fn local_addr(&self) -> SocketAddr {
+        self.stream.local_addr().unwrap()
+    }
+
+    fn remote_addr(&self) -> SocketAddr {
+        self.stream.peer_addr().unwrap()
+    }
+
+    fn open_bidirectional(
+        &self,
+        _stream_id: &StreamId,
+    ) -> Result<(&Self::Read, &Self::Write), Self::Error> {
+        Ok((&self.secret_connection, &self.secret_connection))
```

Comment:
Thinking more about it, the implementation of `MConnection` should take care of the keep-alive semantics and unpack the `PacketMessage` into prefixed streams through `open-bidirectional. Specifically this packet [`Sum`](Thinking more about it, the implementation of `MConnection` should take care of the keep-alive semantics and unpack the `PacketMessage` into prefixed streams through `open-bidi) should be handled in here and the `PacketMsg` case should be unpacked and send to the receivers of a specific stream.

Reply:
> This is not quite what we want. Opening a bidirectional stream means that all bytes read and wrote from and to it will be only for messages of that stream id. In the case of MConn that is symmetric with ChanneldID.

Right, so we'll need to wrap `&self.secret_connection` into something that takes `streamId` into account (i.e. unpacks the message, checks if streamId == thisStreamId and, depending on the result, proxies the message or drops it).

Reply:
> PacketMessage

where I can find it? or it does not exist yet?

Reply:
All the p2p messages are in here https://docs.rs/tendermint-proto/0.18.1/tendermint_proto/p2p/packet/enum.Sum.html

The one which needs to be used for the multiplexing is https://docs.rs/tendermint-proto/0.18.1/tendermint_proto/p2p/struct.PacketMsg.html it has the channel id and the bytes that need to be pass around.